### ISSUE TITLE
How to generate a strategy for a DNN？

### ISSUE THREAD

**Issue by Orion-wyc** [flexflow/FlexFlow#14]
It seems that I haven't figured out how to use **FlexFlow**.
How should a strategy be used in training and how to generate one strategy? 
Can you provide a tutorial? (use the simulator in the scripts?)

In the **strategy** folder, the strategies can not work  and an error occurs (as follow):
'''
strategies.size() = 12
workSpaceSize (1024 MB)
Floating point exception (core dumped)
'''
However, With the default strategy(data parallelism) , it can run successfully. 
'''
......
forwardAlgo(7) time(0.67)
bwdFilterAlgo(5) time(0.77)
bwdDataAlgo(5) time(0.66)
init pool (input): n(64) c(256) h(13) w(13)
init pool (output): n(64) c(256) h(6) w(6)
init linear (input): in_dim(9216) out_dim(4096) batch_size(64)
init linear (input): in_dim(4096) out_dim(4096) batch_size(64)
init linear (input): in_dim(4096) out_dim(1000) batch_size(64)
ELAPSED TIME = 7.2131s, THROUGHPUT = 1135.71 samples/s

'''
Thanks.

**Comment by jiazhihao** [comment]
The strategies are generated by the FlexFlow simulator: https://github.com/flexflow/FlexFlow/blob/master/scripts/simulator.cc. Currently, there is a mismatch between the format of the strategies generated by the FlexFlow and the format expected by the runtime.

I will working on fixing this early next week.

**Comment by Orion-wyc** [comment]
> The strategies are generated by the FlexFlow simulator: https://github.com/flexflow/FlexFlow/blob/master/scripts/simulator.cc. Currently, there is a mismatch between the format of the strategies generated by the FlexFlow and the format expected by the runtime.
> 
> I will working on fixing this early next week.

@jiazhihao Sounds great! Thank you for the reply!

**Comment by knagrecha** [comment]
Has this issue been resolved with the latest updates?

**Comment by zechao-zhang** [comment]
> Has this issue been resolved with the latest updates?

I don't think it works, get reply
"Load parallelization strategy from file ..../FlexFlow/src/runtime/strategy.pb.cc.o
Failed to parse strategy file"

**Comment by Orion-wyc** [comment]
> > Has this issue been resolved with the latest updates?
> 
> I don't think it works, get reply
> "Load parallelization strategy from file ..../FlexFlow/src/runtime/strategy.pb.cc.o
> Failed to parse strategy file"

You can run the latest version of FlexFlow. It is able to parse the old strategy file which is deprecated.
I tried to modifiy the strategy file and load it to the distributed runtime, but get an ASSERTION info.

I hope to solve this problem that troubles me. <qwq>

**Comment by zechao-zhang** [comment]
> > > Has this issue been resolved with the latest updates?
> > 
> > 
> > I don't think it works, get reply
> > "Load parallelization strategy from file ..../FlexFlow/src/runtime/strategy.pb.cc.o
> > Failed to parse strategy file"
> 
> You can run the latest version of FlexFlow. It is able to parse the old strategy file which is deprecated.
> I tried to modifiy the strategy file and load it to the distributed runtime, but get an ASSERTION info.
> 
> I hope to solve this problem that troubles me.

Sorry but where is the old strategy file? I installed FlexFlow just this Tuesday so it is the latest version but only the default strategy DP can be adopted.

**Comment by knagrecha** [comment]
> > > > Has this issue been resolved with the latest updates?
> > > 
> > > 
> > > I don't think it works, get reply
> > > "Load parallelization strategy from file ..../FlexFlow/src/runtime/strategy.pb.cc.o
> > > Failed to parse strategy file"
> > 
> > 
> > You can run the latest version of FlexFlow. It is able to parse the old strategy file which is deprecated.
> > I tried to modifiy the strategy file and load it to the distributed runtime, but get an ASSERTION info.
> > I hope to solve this problem that troubles me.
> 
> Sorry but where is the old strategy file? I installed FlexFlow just this Tuesday so it is the latest version but only the default strategy DP can be adopted.

I found strategy files on the sysmlae branch. Not sure if those are what is being referenced, but I will be testing these to see if they work.

**Comment by jiazhihao** [comment]
@knagrecha Which models you would like to train using FlexFlow? We have updated the interface for describing parallelization strategies, and the strategy files in the sysmlae branches are deprecated.

**Comment by knagrecha** [comment]
> @knagrecha Which models you would like to train using FlexFlow? We have updated the interface for describing parallelization strategies, and the strategy files in the sysmlae branches are deprecated.

@jiazhihao  I'm currently just trying to train AlexNet. Where would I find the updated strategy files?

**Comment by knagrecha** [comment]
I suppose more accurately, my question is how to translate the results of the simulator (if they are still applicable to the current runtime) to a strategy file.

**Comment by Orion-wyc** [comment]
> @knagrecha Which models you would like to train using FlexFlow? We have updated the interface for describing parallelization strategies, and the strategy files in the sysmlae branches are deprecated.

 @jiazhihao  Is that to say that we can use our own strategy to train a DNN but the strategy generated by 'scripts/simulator.cc' should be refactored manually  or by a script to be a xxx_strategy.pb file ?
If we want to train a model with a real dataset (i.e. ImageNet ), does the Dataloader interface in alexnet.cc need to be modified?
The idea of your paper is really interesting, it seems that I haven't figure out the entire usage of the framwork and the accurate meaning of some parameters (i.e. n.dims and dims).

**Comment by ghostplant** [comment]
@bhetherman I compare the Resnet strategies. The optimal strategy is the same as data parallel strategy.

**Comment by Orion-wyc** [comment]
> @bhetherman I compare the Resnet strategies. The optimal strategy is the same as data parallel strategy.

As mentioned in the paper: for ResNet-101, FlexFlow finds strategies similar to data parallelism (except using model parallelism on a single node for the last fully- connected layer) and therefore achieves similar paral- lelization performance. 

Actually, DP performs well on ResNet and other strategies brings little improvement.

So, have you tested strategies with some other DNNs?

**Comment by ghostplant** [comment]
@Orion-wyc The best option strategy I found for Alexnet is to keep conv/pool layers in DP mode, while the following dense layers is executed on a single device. But I don't find some examples that cutting on either C, H, W axis in a NCHW-format tensor will contribute good performance in flexflow.
Anyone who has seen cutting C, H, or W can contribute to the best performance?
